### PR TITLE
avoid non-termination of simplify_exprt::simplify_byte_extract(array_of(...))

### PR DESCRIPTION
**Pull request by kroening**

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 57855b8e732339a6ed2a09f86d40467ceacd12a7).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77779572

File: src/util/simplify_expr.cpp [hunk 1]
```diff
@@ -1734,12 +1734,15 @@ bool simplify_exprt::simplify_byte_extract(byte_extract_exprt &expr)
...
   if(expr.op().id()==ID_array_of &&
-     expr.op().op0().id()==ID_constant)
+     to_array_of_expr(expr.op()).op().id()==ID_constant)
   {
     std::string const_bits=
-      expr2bits(expr.op().op0(),
+      expr2bits(to_array_of_expr(expr.op()).op(),
                 byte_extract_id()==ID_byte_extract_little_endian);
 
+    if(const_bits.empty())
```
Comment (smowton) [inline src/util/simplify_expr.cpp]: Add a comment suggesting why we would have a zero-width constant like this?

File: src/util/simplify_expr.cpp [hunk 2]
```diff
@@ -1734,12 +1734,15 @@ bool simplify_exprt::simplify_byte_extract(byte_extract_exprt &expr)
     return true;
 
   if(expr.op().id()==ID_array_of &&
-     expr.op().op0().id()==ID_constant)
+     to_array_of_expr(expr.op()).op().id()==ID_constant)
```
Comment (smowton) [inline src/util/simplify_expr.cpp:1739]: Suggest the cleanup and the critical fix should be in different commits

File: src/util/simplify_expr.cpp (hunk 1 above)
Comment (tautschnig) [reply src/util/simplify_expr.cpp]: I'm tempted to say that this should be an invariant? (It obviously isn't as that unbounded recursion has been observed, but I'm equally surprised by that.)

**Comment by allredj** [comment]
This fixes https://diffblue.atlassian.net/browse/TG-4174.It needs a regression test, at least on cbmc. I'll write one up.

**Comment by allredj** [comment]
@kroening You can add a regression test by cherry-picking the last commit of this branch: https://github.com/diffblue/cbmc/tree/regression-test-for-2497

File: src/util/simplify_expr.cpp (hunk 1 above)
Comment (kroening) [reply src/util/simplify_expr.cpp]: I think it could return an optional, which, when without value, would signal that there is no bit representation.
Then !const_bits.empty() would be an invariant.

**Comment by kroening** [comment]
Test added, now an invariant

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: fc4aab3ffbf2a9eb26108e31e565958b63240d5c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/78118648

**Review by smowton — APPROVED** [review]
Maybe comment in the test where the failed byte extract comes from?
